Feature/docsite #15 (#16)
zeppelinos/zeppelin-solidity-hot-loader: docs/modules/ROOT/pages/index.adoc

= Solidity Loader

> Solidity Hot Loader is a webpack loader that watches the smart contract `.sol` files you’re using. Once there’s a change, it compiles them, updates the local blockchain using the https://openzeppelin.com/sdk/[OpenZeppelin SDK], and refreshes your app in the browser while preserving your blockchain state. Compare this against the existing Solidity development cycle.

Reviewer: ```suggestion
Solidity Hot Loader is a webpack loader that watches the smart contract `.sol` files you’re using. Once there’s a change, it compiles them, updates the local blockchain using the xref:upgrades::index.adoc[*OpenZeppelin Upgrades*], and refreshes your app in the browser while preserving your blockchain state.
```

> The Solidity Hot Loader development cycle:

Reviewer: I'd swap these two around, I think it looks more impressive that way

Author: Swapped

4.  Restore the state.
5.  Restart dapp to get new contract.

== Overview

=== Installation

[source,console]
----
$ npm install @openzeppelin/solidity-loader json-loader
----

=== Usage

Add the following https://webpack.js.org/configuration/module/#rule[rule] into your webpack configuration:
[source,javascript]
----
solidityLoader: {
  test: /\.sol$/,
  use: [
    { loader: 'json-loader' },
    {
      loader: '@openzeppelin/solidity-loader',
      options: {
        network: 'development',
        disabled: false,
      }
    },
  ],
}
----

NOTE: If you are using `create-react-app`, you may need to either eject the application to modify the webpack configuration, or use https://github.com/timarney/react-app-rewired[`react-app-rewired`].

Import your Solidity contracts into JavaScript like this:
[source,javascript]
----
counterJSON = require('../../contracts/Counter.sol');
----

Changes to your contracts in development will now hot load! ✨

== Learn More

 * Check out xref:getting-started.adoc[Getting Started] if you're new to the Solidity Loader
 * Watch our Devcon V https://youtu.be/KU6_r3sJ2E4[Delightful Developer Experience] presentation, by Igor Yalovoy.

> Igor Yalovoy.

Reviewer: ```suggestion
```

= Solidity Loader

>  * Allows importing `.sol` files directly into source code.

Reviewer: This may be me not being used to webpack loaders, but most of these sound more like a description than useful features. I'd consider removing this list, or trimming it down to only keep high-level features, like the 'tracks dependencies' one, 'automatically recompiles and upgrades in less than x seconds' or 'safe to use: only runs on development networks'.

Author: I removed the list.

 * Allows specifying the contract name for `.sol` files, like `require("Contract.sol?contract=Counter")`.
 * Tracks dependencies for contracts, so children are updated when parents are modified.
 * Converts `.sol` files into `.json` using `oz compile` and `oz update`.
 * Uses the `development` network only, so won't run on any other networks.
 * Set the development network name using the loader's config.
 * Handles race conditions at Webpack pipeline using a lock.
 * Pulls build directory from config.
 * In case `openzeppelin` commands are not available, uses cached versions of `.json` files.
 * Reports errors to console output and dev tools.
 * Compilation can be disabled using the loader's disabled config.

== Overview

> Solidity Hot Loader is a webpack loader that watches the smart contract `.sol` files you’re using. Once there’s a change, it compiles them, updates the local blockchain using the https://openzeppelin.com/sdk/[OpenZeppelin SDK], and refreshes your app in the browser while preserving your blockchain state. Compare this against the existing Solidity development cycle.

Reviewer: In this case, this paragraph (and the lists below) work much better as an intro than a list of features (like in other products). I'd move this to the top of the page.

Author: Moved

The Solidity Hot Loader development cycle:

1.  Make a change to your smart contract `.sol` file and save it.
2.  Wait a bit.
3.  Done.

The existing (slow) Solidity development cycle:

1.  Make a change to your smart contract `.sol` file and save it.
2.  Compile your contract.
3.  Deploy your contract.
4.  Restore the state.
5.  Restart dapp to get new contract.

=== Installation

[source,console]
----

> $ npm install @openzeppelin/solidity-loader

Reviewer: ```suggestion
$ npm install --save-dev @openzeppelin/solidity-loader
```

Author: My understanding is that Solidity loader is used in production for the loading of contract JSON (using .sol name) but the hot part of the loader only works on a development network.  So should be installed as a dependency.

Reviewer: Ah, makes sense. Thanks!

----

=== Usage

> Webpack configuration for Solidity Hot Loader:

Reviewer: I'd change this to a (very) short guide, using verbs:

- add the following webpack configuration
- import your files like this
- everything now works! :sparkles:

Author: Updated usage (having to eject or use something like `react-app-rewired` makes it less simple).

[source,javascript]
----
solidityLoader: {
  test: /\.sol$/,
  use: [
    { loader: 'json-loader' },
    {
      loader: '@openzeppelin/solidity-loader',
      options: {
        network: 'development',
        disabled: false,
      }
    },
  ],
}
----

Example of importing Solidity file into JavaScript:
[source,javascript]
----
counterJSON = require('../../contracts/Counter.sol');
----

== Learn More

>  * Check out xref:getting-started.adoc[Getting Started] with the Tutorial Starter Kit.

Reviewer: ```suggestion
 * Check out xref:getting-started.adoc[Getting Started] if you're new to the Solidity Loader
```

>  * Watch Devcon video https://youtu.be/KU6_r3sJ2E4[Delightful Developer Experience] presented by 

Reviewer: ```suggestion
 * Watch our Devcon V https://youtu.be/KU6_r3sJ2E4[Delightful Developer Experience] presentation, by Igor Yalovoy.
```

Igor Yalovoy.

>  * Finally check out xref:dapp.adoc[Building a dapp with Solidity Hot Loader].

Reviewer: ```suggestion
 * For a step-by-step guide, head to xref:dapp.adoc[Building a Dapp with Solidity Hot Loader].
```
